crates/interpreter/src/valid.rs

```diff
@@ -742,17 +778,21 @@ impl<'a, 'm> Expr<'a, 'm> {
         self.for_each(expected, |x, y| check(x.matches(y)))
     }
 
-    fn push_label(&mut self, type_: FuncType<'m>, kind: LabelKind) -> CheckResult {
+    fn push_label(&mut self, type_: FuncType<'m>, kind: LabelKind<'m>) -> CheckResult {
         self.pops(type_.params)?;
         let stack = type_.params.iter().cloned().map(OpdType::from).collect();
-        let label = Label { type_, kind, polymorphic: false, stack };
+        let label = Label { type_, kind, polymorphic: false, stack, branches: vec![] };
         self.labels.push(label);
         Ok(())
     }
 
     fn end_label(&mut self) -> CheckResult {
+        for branch in core::mem::take(&mut self.label().branches) {
+            self.stitch(&branch);
+        }
+
         let label = self.label();
-        if label.kind == LabelKind::If {
+        if matches!(label.kind, LabelKind::If) {
             check(label.type_.params == label.type_.results)?;
         }
```

Comment:
Here, we want something like this:
```suggestion
        if LabelKind::If(source) = label.kind {
            self.stitch(source, target);
            check(label.type_.params == label.type_.results)?;
        }
```
Where `target` is defined at the beginning of the function with `let target = self.branch();` and reused by all calls to `self.stitch(source, target)`. (The `for` loop will be `for source in take(...)`.)